### PR TITLE
Upgrading to SPDK 24.01

### PR DESCRIPTION
**Pull request by dsavitskiy**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -87,7 +87,7 @@ make
 * ISAL is now mandatory
 * When upgrading from previous versions, clean the SPDK directory
   before building (e.g. git clean).
-* An `AS` is now required, either `yasm` or `nasm`.
+* An `AS` shell var must be set, either `yasm` or `nasm`.
```
Comment (tiagolobocastro) [inline README.md]: Only nasm works now right?

```suggestion
* An `AS` shell var must be set, with `nasm`.
```

File: build.rs [hunk 2]
```diff
@@ -52,18 +52,28 @@ fn get_target_dir() -> PathBuf {
     p
 }
 
+/// Returns nightly rustfmt path, if it can be found. Returns None oetherwise.
```
Comment (tiagolobocastro) [inline build.rs]:
```suggestion
/// Returns nightly rustfmt path, if it can be found. Returns None otherwise.
```

File: build.rs [hunk 3]
```diff
@@ -52,18 +52,28 @@ fn get_target_dir() -> PathBuf {
...
+fn rust_fmt_nightly() -> Option<PathBuf> {
+    let mut cmd = std::process::Command::new("rustup");
```
Comment (tiagolobocastro) [inline build.rs]: I wonder, should we find the nightly from the nix environment itself? Ex: I don't keep a rustup nightly myself

File: nix/pkgs/libspdk/default.nix [hunk 4]
```diff
@@ -1,86 +1,124 @@
-{ binutils
-, cunit
+{ pkgs
+
+, build-type
+, with-fio
+, multi-outputs ? false
```
Comment (tiagolobocastro) [inline nix/pkgs/libspdk/default.nix:5]: since we are making these changes can just always use multi-outputs?

File: nix/shell/rust.nix [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+    llvmPackages.clang
+    llvmPackages.libclang
+    pkg-config
+    procps
+    utillinux
+  ];
+
+  shellEnv = {
+    # Path to nightly Rust, needed for linter and style.
+    RUST_NIGHTLY_PATH = rustChannels.nightly;
```
Comment (tiagolobocastro) [inline nix/shell/rust.nix:21]: Ah, can't the build.rs find this one?

File: wrapper.h [hunk 6]
```diff
@@ -1,3 +1,10 @@
+
+#include <spdk/assert.h>
+
+#ifndef assert
```
Comment (tiagolobocastro) [inline wrapper.h]: Could we include the "common" here rather than "harcode" again?

**Review by tiagolobocastro — APPROVED** [review]

File: build_helpers/library_config.rs [hunk 7]
```diff
@@ -319,27 +320,63 @@ impl LibraryConfig {
     ) -> Result<(), Error> {
         let lib_name = OsString::from(&lib_name);
         let library = self.cfg.probe(lib_name.to_str().unwrap())?;
+
+        println!("Found pkg config dependencies for {lib_name:?}:");
+
         for name in &library.libs {
             let lib = Library::new(name, &lib_name, self);
+            println!("    link lib {n:?}", n = lib.pkg_name);
```
Comment (dsharma-dc) [inline build_helpers/library_config.rs:328]: Extra white spaces can be removed from `println` in this file ( unless intentionally kept).

**Review by dsharma-dc — APPROVED** [review]

File: README.md (hunk 1 above)
Comment (dsavitskiy) [reply README.md]: I'm going to rewrite the README.

File: build.rs (hunk 2 above)
Comment (dsavitskiy) [reply build.rs]: will fix

File: build.rs (hunk 3 above)
Comment (dsavitskiy) [reply build.rs]: Nix already sets up nightly for our Git commit hook's fmt purposes.
And we have a shell var set up by Nix, 'RUST_NIGHTLY_PATH'. We can use, surely. I'll fix.
But I'd keep this code as well, as a fallback.

File: nix/shell/rust.nix (hunk 5 above)
Comment (dsavitskiy) [reply nix/shell/rust.nix:21]: Yes it can. See my comment above.

File: wrapper.h (hunk 6 above)
Comment (dsavitskiy) [reply wrapper.h]: Yes indeed, I will fix.

File: build_helpers/library_config.rs (hunk 7 above)
Comment (dsavitskiy) [reply build_helpers/library_config.rs:328]: It's an intention indent, to make the output a bit easier to follow. The header for this section is printed by this func, in the beginning.

**Comment by dsavitskiy** [comment]
bors try

File: nix/pkgs/libspdk/default.nix (hunk 4 above)
Comment (dsavitskiy) [reply nix/pkgs/libspdk/default.nix:5]: This may break release build scripts. We can do this as the next iteration.

**Comment by dsavitskiy** [comment]
bors merge

**Comment by dsavitskiy** [comment]
bors merge

**Pull request merged**